libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,634 @@
+  s_store.tx_event = settings->tx_event;
+  s_store.rx_event = settings->rx_event;
+  s_store.repeat_event = settings->repeat_event;
+  s_store.error_event = settings->error_event;
+
+  fifo_init(&s_fifo, s_buffer);
+  prv_init_fsm((void *)&s_store);
+  s_store.status = DATAGRAM_STATUS_IDLE;
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_datagram_rx(uint8_t *data, size_t len, bool start_message) {
+  if (len > CAN_BUFFER_SIZE) {
+    return STATUS_CODE_OUT_OF_RANGE;
+  }
+
+  // Only run if start_listener called
+  if (!s_store.rx_listener_enabled) {
+    return STATUS_CODE_UNINITIALIZED;
+  }
+  if (start_message) {
+    // Only process start messages if rx is not currently active
+    // or if soft error has occured
+    if (s_store.status != DATAGRAM_STATUS_ACTIVE || s_store.soft_error_flag) {
+      // Start frame rcv'd -> status now active
+      // Commence rx side of state machine
+      s_store.status = DATAGRAM_STATUS_ACTIVE;
+      s_store.soft_error_flag = false;
+      event_raise_no_data(s_store.rx_event);
+      watchdog_start(&s_watchdog, RX_WATCHDOG_TIMEOUT_MS, prv_rx_watchdog_expiry_cb, NULL);
+      return STATUS_CODE_OK;
+    } else {
+      return STATUS_CODE_UNREACHABLE;
+    }
+  } else if (s_store.status != DATAGRAM_STATUS_ACTIVE) {
+    // If start message not received, do not rx data
+    return STATUS_CODE_UNINITIALIZED;
+  }
+
+  // Rx normal data
+  watchdog_kick(&s_watchdog);
+  prv_write_rx_buffer(data, len);
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_datagram_start_listener(CanDatagramRxConfig *config) {
+  // Set up rx fifo for rxed datagram messages, zero data buffers
+  if (s_store.status == DATAGRAM_STATUS_ACTIVE) {
+    return STATUS_CODE_UNREACHABLE;
+  }
+  // Set up data buffers
+  CanDatagram *dgram = &s_store.dgram;
+  dgram->data = config->data;
+  dgram->destination_nodes = config->destination_nodes;
+
+  // Store config * to be filled out after rx complete
+  s_store.rx_info = config;
+
+  // Store current node id
+  s_store.node_id = config->node_id;
+
+  s_store.rx_listener_enabled = true;
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_datagram_start_tx(CanDatagramTxConfig *config) {
+  // Verify datagram in idle state
+  if (s_store.status == DATAGRAM_STATUS_ACTIVE) {
+    return STATUS_CODE_UNREACHABLE;
+  }
+  // Do error checking before commencing
+  if (config->tx_cb == NULL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  if (config->data_len > MAX_DATA_SIZE_BYTES) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  // Datagram initialization
+  CanDatagram *dgram = &s_store.dgram;
+  dgram->protocol_version = CAN_DATAGRAM_VERSION;
+  dgram->dgram_type = config->dgram_type;
+  dgram->destination_nodes_len = config->destination_nodes_len;
+  dgram->data_len = config->data_len;
+  dgram->destination_nodes = config->destination_nodes;
+  dgram->data = config->data;
+  dgram->crc = prv_can_datagram_compute_crc();
+
+  s_store.tx_cb = config->tx_cb;
+  s_store.tx_cb(NULL, 0, true);
```

Comment:
Hm, I suppose this does conform to the standard - my original thought was that the start message would also contain data, so we'd just set the start flag to true on the first tx_cb call. That way is more efficient especially for short messages, but this is simpler - your call on whether to change it or not.